mne/decoding/gsoc.py

```diff
@@ -0,0 +1,523 @@
+
+    .. note:: This does not handle overlapping events. Use original
+              preprocessing.xdawn instead.
+
+    Parameters
+    ----------
+    n_components : int (default 2)
+        The number of components to decompose M/EEG signals.
+    n_chan : int
+         Integer indicating the number of channels.
+    signal_cov : None | Covariance | ndarray, shape (n_channels, n_channels)
+        (default None). The signal covariance used for whitening of the data.
+        if None, the covariance is estimated from the epochs signal.
+    reg : float | str | None (default None)
+        if not None, allow regularization for covariance estimation
+        if float, shrinkage covariance is used (0 <= shrinkage <= 1).
+        if str, optimal shrinkage using Ledoit-Wolf Shrinkage ('ledoit_wolf')
+        or Oracle Approximating Shrinkage ('oas').
+
+    Attributes
+    ----------
+    filters_ : dict of ndarray
+        If fit, the Xdawn components used to decompose the data for each event
+        type, else empty.
+    patterns_ : dict of ndarray
+        If fit, the Xdawn patterns used to restore M/EEG signals for each event
+        type, else empty.
+
+
+    See Also
+    --------
+    mne.preprocessing.Xdawn
+    CSP
+
+    References
+    ----------
+    [1] Rivet, B., Souloumiac, A., Attina, V., & Gibert, G. (2009). xDAWN
+    algorithm to enhance evoked potentials: application to brain-computer
+    interface. Biomedical Engineering, IEEE Transactions on, 56(8), 2035-2043.
+
+    [2] Rivet, B., Cecotti, H., Souloumiac, A., Maby, E., & Mattout, J. (2011,
+    August). Theoretical analysis of xDAWN algorithm: application to an
+    efficient sensor selection in a P300 BCI. In Signal Processing Conference,
+    2011 19th European (pp. 1382-1386). IEEE.
+    """
+
+    def __init__(self, n_components=2, n_chan=None, signal_cov=None,
+                 reg=None):
+        """init xdawn."""
+        if n_chan is None:
+            raise ValueError("n_chan cannot be none. Please provide the "
+                             "number of channels")
+        self.n_chan = n_chan
+        self.n_components = n_components
+        self.signal_cov = signal_cov
+        self.reg = reg
+
+    def fit(self, X, y):
+        """Fit Xdawn from epochs.
+
+        Parameters
+        ----------
+        X : ndarray, shape(n_channels, n_times * n_freq)
+            Data of epochs.
+        y : ndarray shape(n_samples,)
+            Target values.
+
+        Returns
+        -------
+        self : XdawnTransformer instance
+            The XdawnTransformer instance.
+        """
+        from sklearn.preprocessing import LabelEncoder
+
+        if X.ndim != 2 or not isinstance(X, np.ndarray):
+            raise ValueError("X should be 2 dimensional ndarray")
+        if not isinstance(y, np.ndarray):
+            raise ValueError("Labels must be numpy array")
+
+        epochs_data = X.reshape(X.shape[0], self.n_chan, X.shape[1] /
```

Comment:
instead of using X.shape[0], X.shape[1], do `n_epochs, n_chans, n_times = X.shape` it's more explicit